go-controller/pkg/ovn/obj_retry.go

```diff
@@ -1134,31 +1181,19 @@ func (oc *Controller) deleteResource(objectsToRetry *retryObjs, obj, cachedObj i
 	}
 }
 
-type localRetryEntry struct {
-	key string // the key in the retryObjs map holding retryObjs value
-	// cached newObj holds k8s resource failed during add operation
-	newObj interface{}
-	// cached oldObj holds k8s resource failed during delete operation
-	oldObj interface{}
-	// resource retrieved from the K8s API Server
-	kObj interface{}
-}
-
-func (oc *Controller) resourceRetry(r *retryObjs, lre *localRetryEntry, now time.Time) {
-	objKey := lre.key
-	entry := r.ensureRetryEntryLocked(objKey, nil)
-	if entry == nil {
+func (oc *Controller) resourceRetry(r *retryObjs, objKey string, now time.Time) {
+	lockedKey := r.lockKey(objKey)
+	defer r.unlockKey(lockedKey)
+	entry, loaded := r.getRetryObj(lockedKey)
```

Comment:
There is a relationship between keyLocks and entries, where no one should modify an entry without a lock for the same key in keyLocks map. However, this is not enforced. I think we should create a common ExecuteWithLock(key, func) that will lock the key and execute whatever on the same keyed entry. In the future once this is moved to its own library, a user will not have access to the entries map directly. I'm worried about bugs where a developer accidentally modifies an entry without a key lock.

Reply:
Actually it is enforced (that's why we added `type LockedKey string`): entries are not accessed directly by anything except SyncCache, and all SyncCache methods that access entries look like
```
func (c *SyncCache) Load(key LockedKey) (value *retryObjEntry, loaded bool)
func (c *SyncCache) LoadOrStore(key LockedKey, newEntry *retryObjEntry) *retryObjEntry
func (c *SyncCache) Delete(key LockedKey)
```
they have `LockedKey` as an argument and you can only get it by calling 
```
func (c *SyncCache) LockKey(key string) LockedKey
```
So the usual workflow will be like
```
lockedKey := oc.retrySmth.lockKey(key)
defer oc.retrySmth.unlockKey(lockedKey)
// do something like
oc.retrySmth.initRetryObjWithDelete(np.policy, lockedKey, np, false)
isLastPolicyInNamespace := len(nsInfo.networkPolicies) == 1
if err := oc.destroyNetworkPolicy(np, isLastPolicyInNamespace); err != nil {
	klog.Errorf("Failed to delete network policy: %s, error: %v", key, err)
} else {
	oc.retryNetworkPolicies.deleteRetryObj(lockedKey)
	delete(nsInfo.networkPolicies, np.name)
}

```

Reply:
Right, LockedKey is something the SyncCache exports to SyncCache users, but that does not meant retryObjs should export the same thing to retryObjs users as happened in the last push. But this is all a bit confusing since we are throwing everything in with no clear drawn line between interfaces.

I would not leak the concept of locked key outside of objs_retry.go, SyncCache and SyncCache direct users until we put some order here.

Reply:
I think it is true - if every key handling includes only 1 operation with this key, all key locking interface can be hidden, but that is not true for retry_obj algorithms - they get, create, update, delete one key just during update event hendling